contentprovider.go

```diff
@@ -212,6 +212,31 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch) []LineMatch {
 		}
 		finalMatch.Line = data[lineStart:lineEnd]
 
+		if numContextLines > 0 {
+			var before [][]byte
+			var after [][]byte
+			lineNum := 0
+			prev := -1
+			for lineNum < num+numContextLines && prev < len(data) {
+				next := bytes.IndexByte(data[prev+1:], '\n')
+				if next == -1 {
+					next = len(data)
+				} else {
+					next += prev + 1
+				}
+				lineNum += 1
+				if (lineNum < num) && (num-lineNum <= numContextLines) {
+					before = append(before, data[prev+1:next])
+				}
+				if (lineNum > num) && (lineNum-num <= numContextLines) {
+					after = append(after, data[prev+1:next])
+				}
+				prev = next
+			}
```

Comment:
this code seems very inefficient since you always start at the beginning of the file. I recommend two possible alternatives:
- have a frontier that goes backwards numContextLines times using `bytes.LastIndexByte` to calculate the before lines. Then the reverse of that for going forward. Start at lineStart and lineEnd respectively
- You can probably use `p.newlines` above (line 168) for faster operations. However, I assume you usually have low numContextLines so the first approach is likely better.

Reply:
Thanks, using `p.newlines` turned out to be more readable IMO